pallets/parachain-system/src/lib.rs

```diff
@@ -960,6 +1055,68 @@ impl<T: Config> Pallet<T> {
 		weight_used
 	}
 
+	/// Drop blocks from the unincluded segment with respect to the latest parachain head.
+	///
+	/// No-op if [`MaxUnincludedLen`] is not set.
+	fn maybe_drop_included_ancestors(relay_state_proof: &RelayChainStateProof) -> Weight {
+		let mut weight_used = Weight::zero();
+		// If `MaxUnincludedLen` is present in the storage, parachain head
+		// is always expected to be included into the relay storage proof.
+		let para_head_with_len = <MaxUnincludedLen<T>>::get().map(|max_len| {
+			(
+				relay_state_proof
+					.read_included_para_head()
+					.expect("Invalid para head in relay chain state proof"),
+				max_len,
+			)
+		});
+		weight_used += T::DbWeight::get().reads(1);
+		let Some((para_head, max_len)) = para_head_with_len else { return weight_used };
+
+		let para_head_hash = T::Hashing::hash(&para_head.0);
+		if !max_len.is_zero() {
+			let (dropped, left_count): (Vec<Ancestor<T::Hash>>, u32) =
+				<UnincludedSegment<T>>::mutate(|chain| {
+					// Drop everything up to the block with an included para head, if present.
+					let idx = chain
+						.iter()
+						.position(|block| {
+							let head_hash = block.para_head_hash().expect(
+								"para head hash is updated during block initialization; qed",
+							);
+							head_hash == &para_head_hash
+						})
+						.map_or(0, |idx| idx + 1); // inclusive.
+
+					let left_count = (idx..chain.len()).count() as u32;
+					let dropped = chain.drain(..idx).collect();
+					(dropped, left_count)
+				});
+			weight_used += T::DbWeight::get().reads_writes(1, 1);
+
+			// sanity-check there's place for the block at finalization phase.
+			//
+			// TODO: this potentially restricts parachains from decreasing `MaxUnincludedLen` value.
```

Comment:
I think the answer here, now that we've moved this to run in the inherent, is "it doesn't".

`on_initialize` always runs before inherents, so any pallet setting the `MaxUnincludedLen` in code that executes in the `on_initialize` of any pallet would run before this, so it is safe.

Reply:
What about scenario when there're 4 blocks in the unincluded segment, and the len is reduced to e.g. 1?

Reply:
That is fine. For example, assume that there is some unconditional code that sets the maximum length to 1 in `on_initialize` at a certain block height. For more dynamically controlled systems, the reasoning is the same but slightly more complicated.

As long as the unincluded segment has length > 1, we can't author a new block. But new relay-parents will "drain" the unincluded segment from the back, so eventually the unincluded segment will be short enough to push a new block onto. This is the desired behavior of shrinking the maximum segment length.

Reply:
It's rather that we are creating a block validity rule based on the relay-parent. It'll be impossible to create valid blocks on top of some `(para parent, relay parent)` pairs, and that failure will manifest in the inherent that handles the storage proof. Anything that runs in `on_initialize` could alter the block validity condition (e.g. altering the maximum unincluded segment length). We will probably also want hooks for people to write dynamic block validity rules based on the relay-parent in question.